src/passes/J2clOpts.cpp

```diff
@@ -0,0 +1,210 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Optimize J2CL specifics construct to simplify them and enable further
+// optimizations by other passes.
+//
+
+#include "ir/global-utils.h"
+#include "ir/utils.h"
+#include "opt-utils.h"
+#include "pass.h"
+#include "passes.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+static bool isOnceFunction(Function* f) {
+  return f->name.hasSubstring("_@once@_");
+}
+
+using AssignmentCountMap = std::unordered_map<Name, Index>;
+
+// TODO: parallel
+class GlobalAssignmentCollector
+  : public WalkerPass<PostWalker<GlobalAssignmentCollector>> {
+public:
+  GlobalAssignmentCollector(AssignmentCountMap& assignmentCounts)
+    : assignmentCounts(assignmentCounts) {}
+
+  void visitGlobalSet(GlobalSet* curr) {
+    // Avoid optimizing class initialization condition variable itself.
+    if (curr->name.startsWith("f_$initialized__")) {
+      return;
+    }
+    assignmentCounts[curr->name]++;
+  }
+
+private:
+  AssignmentCountMap& assignmentCounts;
+};
+
+// TODO: parallel
+class ConstantHoister : public WalkerPass<PostWalker<ConstantHoister>> {
+public:
+  ConstantHoister(AssignmentCountMap& assignmentCounts)
+    : assignmentCounts(assignmentCounts) {}
+
+  int optimized = 0;
+
+  void visitFunction(Function* curr) {
+    if (!isOnceFunction(curr)) {
+      return;
+    }
+    int optimizedBefore = optimized;
+    if (auto* block = curr->body->dynCast<Block>()) {
+      for (auto*& expr : block->list) {
+        maybeHoistConstant(expr);
+      }
+    } else {
+      maybeHoistConstant(curr->body);
+    }
+
+    if (optimized != optimizedBefore) {
+      // We introduced "nop" instruction. Run the vacuum to cleanup.
+      // TODO: maybe we should not introduce "nop" in the first place and try
+      // removing instead.
+      PassRunner runner(getModule());
+      runner.add("vacuum");
+      runner.setIsNested(true);
+      runner.runOnFunction(curr);
+    }
+  }
+
+private:
+  void maybeHoistConstant(Expression* expr) {
+    auto set = expr->dynCast<GlobalSet>();
+    if (!set) {
+      return;
+    }
+
+    if (assignmentCounts[set->name] != 1) {
+      // The global assigned in multiple places, so it is not safe to
+      // hoist them as global constants.
+      return;
+    }
+
+    if (!GlobalUtils::canInitializeGlobal(*getModule(), set->value)) {
+      // It is not a valid constant expression so cannot be hoisted to
+      // global init.
```

Comment:
It is convenient that this rules out all the dangerous situations, like local state, doing a call, etc.

Reply:
Exactly, it was nicely fit!